File: obfuscate.py
import os
import time
import json
import re
import shutil

OUTPUT_DIR = os.path.join(os.path.dirname(__file__), 'obfuscated')
EXCLUDE_LIST = ['node_modules', '.git', 'dist', 'obfuscated', 'obfuscate.js', 'obfuscate.py', '.vscode', '.github', 'server.js', 'server.py', 'requirements.txt', 'venv', '.env']
VERSION_TOKEN = str(int(time.time() * 1000))

def minify_html(content):
    content = re.sub(r'<!--[\s\S]*?-->', '', content)
    content = re.sub(r'>\s+<', '><', content)
    content = re.sub(r'\s{2,}', ' ', content)
    content = re.sub(r'src="([^"]+\.(js|png|jpg|jpeg|svg|webp))"', f'src="\\1?v={VERSION_TOKEN}"', content)
    content = re.sub(r'href="([^"]+\.(css|png|jpg|jpeg|svg|webp))"', f'href="\\1?v={VERSION_TOKEN}"', content)
    return content.strip()

def minify_css(content):
    content = re.sub(r'\/\*[\s\S]*?\*\/', '', content)
    content = re.sub(r'\s+', ' ', content)
    content = re.sub(r'\s*([{}:;,])\s*', r'\1', content)
    return content.strip()

def obfuscate_js(content):
    processed = re.sub(r'\/\*[\s\S]*?\*\/', '', content)
    lines = processed.split('\n')
    new_lines = []
    for line in lines:
        comment_idx = line.find('//')
        if comment_idx != -1:
            if comment_idx == 0 or line[comment_idx-1] != ':':
                line = line[:comment_idx]
        new_lines.append(line)
    processed = ' '.join(new_lines)
    processed = re.sub(r'\s+', ' ', processed).strip()
    
    mangle_map = {
        'navItems': '_0x3a', 'viewSections': '_0x4a', 'scrollReveal': '_0x7a',
        'createBackgroundParticles': '_0x8a', 'initCustomCursor': '_0x5a',
        'renderAll': '_0x6a', 'startApp': '_0x9a', 'startAdmin': '_0xba'
    }
    for key, val in mangle_map.items():
        processed = re.sub(fr'\b{key}\b', val, processed)
        
    export_targets = ['initCustomCursor', 'renderAll', 'startApp', 'startAdmin', 'deleteExecutor', 'deleteGame', 'sha256']
    window_exports = ''
    for target in export_targets:
        if target in content:
            actual = mangle_map.get(target, target)
            window_exports += f"window['{target}'] = {actual};"
            
    return f"{processed};{window_exports}".replace('\n', '').strip()

def process_directory(src, dest):
    if not os.path.exists(dest): os.makedirs(dest)
    for item in os.listdir(src):
        if item in EXCLUDE_LIST: continue
        src_path = os.path.join(src, item)
        dest_path = os.path.join(dest, item)
        
        if os.path.isdir(src_path):
            process_directory(src_path, dest_path)
        else:
            ext = os.path.splitext(item)[1].lower()
            try:
                if ext == '.html':
                    with open(src_path, 'r', encoding='utf-8') as f:
                        with open(dest_path, 'w', encoding='utf-8') as df:
                            df.write(minify_html(f.read()))
                elif ext == '.css':
                    with open(src_path, 'r', encoding='utf-8') as f:
                        with open(dest_path, 'w', encoding='utf-8') as df:
                            df.write(minify_css(f.read()))
                elif ext == '.js':
                    with open(src_path, 'r', encoding='utf-8') as f:
                        with open(dest_path, 'w', encoding='utf-8') as df:
                            df.write(obfuscate_js(f.read()))
                else:
                    shutil.copy2(src_path, dest_path)
            except:
                shutil.copy2(src_path, dest_path)

if __name__ == '__main__':
    if os.path.exists(OUTPUT_DIR):
        for item in os.listdir(OUTPUT_DIR):
            if item == '.git': continue
            path = os.path.join(OUTPUT_DIR, item)
            if os.path.isdir(path): shutil.rmtree(path)
            else: os.remove(path)
    else:
        os.makedirs(OUTPUT_DIR)
        
    process_directory(os.path.dirname(__file__), OUTPUT_DIR)
    print('Obfuscation complete.')

File: obfuscate.js
const fs = require('fs');
const path = require('path');

const OUTPUT_DIR = path.join(__dirname, 'obfuscated');
const EXCLUDE_LIST = ['node_modules', '.git', 'dist', 'obfuscated', 'obfuscate.js', 'obfuscate.py', '.vscode', '.github'];

const VERSION_TOKEN = Date.now();

function minifyHTML(content) {
    let minified = content
        .replace(/<!--[\s\S]*?-->/g, '')
        .replace(/>\s+</g, '><')
        .replace(/\s{2,}/g, ' ')
        .trim();

    // Cache Busting: Add version to scripts and links
    minified = minified.replace(/src="([^"]+\.(js|png|jpg|jpeg|svg|webp))"/g, `src="$1?v=${VERSION_TOKEN}"`);
    minified = minified.replace(/href="([^"]+\.(css|png|jpg|jpeg|svg|webp))"/g, `href="$1?v=${VERSION_TOKEN}"`);

    return minified;
}

function minifyCSS(content) {
    return content
        .replace(/\/\*[\s\S]*?\*\//g, '')
        .replace(/\s+/g, ' ')
        .replace(/\s*([{}:;,])\s*/g, '$1')
        .trim();
}

function obfuscateJS(content) {
    // 1. Minification (Safe comment removal)
    let processed = content
        .replace(/\/\*[\s\S]*?\*\//g, '') // Multi-line
        .split('\n')
        .map(line => {
            const commentIdx = line.indexOf('//');
            if (commentIdx !== -1 && line[commentIdx - 1] !== ':') {
                return line.slice(0, commentIdx);
            }
            return line;
        })
        .join(' ')
        .replace(/\s+/g, ' ')
        .trim();

    // 3. Selective Variable Mangling (Excluding known globals and key functions)
    const mangleMap = {
        'navItems': '_0x3a',
        'viewSections': '_0x4a', 'scrollReveal': '_0x7a', 'createBackgroundParticles': '_0x8a',
        'initCustomCursor': '_0x5a', 'renderAll': '_0x6a', 'startApp': '_0x9a', 'startAdmin': '_0xba'
    };
    for (const [key, val] of Object.entries(mangleMap)) {
        processed = processed.replace(new RegExp(`\\b${key}\\b`, 'g'), val);
    }

    // 4. Ensure global accessibility for key functions
    let windowExports = '';
    const exportTargets = {
        'initCustomCursor': '_0x5a',
        'renderAll': '_0x6a',
        'startApp': '_0x9a',
        'startAdmin': '_0xba',
        'deleteExecutor': 'deleteExecutor',
        'deleteGame': 'deleteGame',
        'sha256': 'sha256'
    };

    for (const [key, val] of Object.entries(exportTargets)) {
        if (content.includes(key)) {
            windowExports += `window['${key}'] = ${val};`;
        }
    }

    return `${processed};${windowExports}`.replace(/\n/g, '').trim();
}

function processDirectory(src, dest) {
    if (!fs.existsSync(dest)) fs.mkdirSync(dest, { recursive: true });
    const entries = fs.readdirSync(src, { withFileTypes: true });
    for (const entry of entries) {
        if (EXCLUDE_LIST.includes(entry.name)) continue;
        const srcPath = path.join(src, entry.name);
        const destPath = path.join(dest, entry.name);
        if (entry.isDirectory()) {
            processDirectory(srcPath, destPath);
        } else {
            const ext = path.extname(entry.name).toLowerCase();
            try {
                let content = fs.readFileSync(srcPath, 'utf8');
                if (ext === '.html') fs.writeFileSync(destPath, minifyHTML(content));
                else if (ext === '.css') fs.writeFileSync(destPath, minifyCSS(content));
                else if (ext === '.js') fs.writeFileSync(destPath, obfuscateJS(content));
                else fs.copyFileSync(srcPath, destPath);
            } catch (err) {
                fs.copyFileSync(srcPath, destPath);
            }
        }
    }
}

if (fs.existsSync(OUTPUT_DIR)) {
    const items = fs.readdirSync(OUTPUT_DIR);
    for (const item of items) {
        if (item === '.git') continue;
        const itemPath = path.join(OUTPUT_DIR, item);
        if (fs.statSync(itemPath).isDirectory()) fs.rmSync(itemPath, { recursive: true, force: true });
        else fs.unlinkSync(itemPath);
    }
} else {
    fs.mkdirSync(OUTPUT_DIR, { recursive: true });
}

processDirectory(__dirname, OUTPUT_DIR);
console.log('Obfuscation complete.');
